gui/default/syncthing/folder/ignoresService.js

```diff
@@ -0,0 +1,207 @@
+                self.data.text = response.map(function(r) { return r.text; }).join('\n');
+                self.data.patterns = response;
+                self.data.disabled = false;
+                return self.data;
+            }).catch(function (err) {
+                self.data.text = '';
+                throw err;
+            });
+        };
+
+        self.save = function(folderId, ignores) {
+            return $http.post('rest/db/ignores',
+                { ignore: ignores },
+                { params: { folder: folderId } }
+            );
+        };
+
+        self.parseText = function() {
+            if (self.data.text.length === 0) {
+                self.data.patterns = [];
+            } else {
+                self.data.patterns = self.data.text
+                    .split('\n')
+                    .map(parsePattern);
+            }
+            return self.data.patterns;
+        };
+
+        self.addPattern = function(text) {
+            var newPattern = parsePattern(text);
+            var afterIndex = findLastIndex(self.data.patterns, function(pattern) {
+                return pattern.isSimple && newPattern.isSimple && newPattern.matchFunc(pattern.path);
+            });
+            self.data.patterns.splice(afterIndex + 1, 0, newPattern);
+            self.data.text = self.data.patterns.map(function(r) { return r.text; }).join('\n');
+        };
+
+        self.removePattern = function(text) {
+            var index = self.data.patterns.findIndex(function(pattern) {
+                return pattern.text === text;
+            });
+            if (index >= 0) {
+                self.data.patterns.splice(index, 1);
+                self.data.text = self.data.patterns.map(function(r) { return r.text; }).join('\n');
+            }
+        };
+
+        self.matchingPattern = function(file) {
+            return self.data.patterns.find(function(pattern) {
+                // Only consider patterns that match a simple path
+                if (!pattern.isSimple) return false;
+
+                var absPath = '/' + file.path;
+                return pattern.matchFunc(absPath);
+            });
+        }
+
+        /*
+         * private definitions
+         */
+
+        function getIgnores(folderId) {
+            return $http.get('rest/db/ignores', {
+                params: { folder: folderId }
+            }).then(function (response) {
+                var data = response.data;
+                return (data.ignore || []).map(parsePattern);
+            });
+        }
+
+        function parsePattern(line) {
+            var stripResult = stripPrefix(line.trim());
+            var prefixes = stripResult[0];
+            var hasPrefix = prefixes['(?i)'] || prefixes['(?d)'];
+            var path = toPath(stripResult[1]);
+            var matchFunc = !hasPrefix && matcher(path);
+
+            return {
+                text: line,
+                isSimple: !!matchFunc,
```

Comment:
Again, unfamiliar with js, is this a not not, or not null?

Reply:
This is a not not. Coercing the absence/presence of matchFunc into a boolean.